### DISCUSSION TITLE
Accessing NetworkX nodes when using custom classes as nodes

### DISCUSSION THREAD

**Discussion by AlvaroJoseLopes** [networkx/networkx#6839] [Q&A] [answered]
Hey Guys,

I'm working on a heterogeneous network for modeling a dataset for Recommender Systems and I was thinking about using a custom class to represent each type of node, for example: `ItemNode`, `UserNode` and `PropertyNode`. However, I encountered some issues when trying to access a specific node when using a copy of the same instance or its hash value. 

It appears that NetworkX requires using the exact same instance of the class as the key to access nodes successfully, otherwise, it raises a **KeyError**.

Here's a code snippet that demonstrates the issue:
```python
import networkx as nx

class CustomNode:
    def __init__(self, value):
        self.value = value
        # ...
    
    def __str__(self):
        return f'CustomNode {self.value}'
    
    def __repr__(self):
        return f'CustomNode({self.value})'

    def __hash__(self):
        return hash(str(self))

G = nx.Graph()

node1 = CustomNode(1)
node2 = CustomNode(2)

G.add_node(node1)
G.add_node(node2)

G.add_edge(node1, node2)

print("Nodes:", G.nodes())
print("Edges:", G.edges())

# Trying to access nodes
print('Node 1:', G[node1]) # Works
print('Node 1: ', G[hash('CustomNode 1')]) # Doesnt work
node1_copy = CustomNode(1)
print(G[node1_copy]) # also doesnt work
```
Is this the expected behavior in NetworkX? I was hoping to access each node without having to use the same instance that I used when adding the node. Is there any other way to achieve this?

I'm also open to suggestions on a better way to model this problem in NetworkX. I'd greatly appreciate any insights or advice on how to handle this situation effectively.

**Comment by dschult** [comment, marked as answer]
Looks like you need to implement the `__eq__` method in your customized class. 
NetworkX uses dicts under the hood, so if you can get it working for a dict (add the custom class instance to a dict, then try to look up that instance using another newly created instance) then it should work with NetworkX.
dicts use __hash__ to find a location in the dict to store the key, but checks using `__eq__` to see if there are any items that have the same hash but aren't equal.  In your case it could look something like:
```python
def __eq__(self, other):
    if isinstance(self, CustomNode) and isinstance(other, CutsomNode):
        return self.value == other.value
    return False
```

You might consider using `1` and `2` as your nodes and then storing the node type as a node attribute:
```python
G.add_node(1, node_type = "User")
G.add_node(2, node_type = "Property")
```
This requires unique identifiers for the nodes across all types, but seems simpler than creating an custom instance for each node...  But both methods should work well and your application may have reasons for using the custom instances.

**Reply by AlvaroJoseLopes** [reply]
I see.... It worked, Thanks
I'll take a closer look at both options to figure out which one works best for my application.
Thanks again
